wiring/src/spark_wiring_ble.cpp

```diff
@@ -1834,14 +1834,14 @@ int BlePeerDevice::connect(const BleAddress& addr, const BleConnectionParams* pa
     connCfg.conn_params = params;
     connCfg.callback = BleLocalDevice::getInstance().impl()->onBleLinkEvents;
     connCfg.context = BleLocalDevice::getInstance().impl();
-    int ret = hal_ble_gap_connect(&connCfg, &impl()->connHandle(), nullptr);
-    if (ret != SYSTEM_ERROR_NONE) {
-        impl()->connHandle() = BLE_INVALID_CONN_HANDLE;
-        return ret;
-    }
-    bind(addr);
     {
         WiringBleLock lk;
+        int ret = hal_ble_gap_connect(&connCfg, &impl()->connHandle(), nullptr);
```

Comment:
We should revert this change, since we have increased the BLE event thread priority to avoid some race conditions. The change here may cause dead lock on Gen3 if there is asynchronous BLE event generated and to be propagated prior to connected event.

Reply:
I recall that this change was going to fix the race condition that `hal_ble_gap_connect()` has exited, however, the peer device hasn't been added to wiring cache and a new BLE event is propagated, which results in peer device not found in the wiring event callback. So a new fix is to propagate the connected event to wiring layer followed by adding the peer device to cache, but not notifying (or maybe we can?) user application if user initiate the connection establishedment. @avtolstoy

Reply:
Can you elaborate on the

> but not notifying (or maybe we can?) user application if user initiate the connection establishedment.

?

Reply:
Currently the `BLE.connect()` API is designed to not generate connected event for application hook, because this API is a blocking API. So to iron out the race condition as I mentioned above, we can change the current scheme.

Reply:
Okay, I see. What are the other events that are propagated from HAL to wiring right after `hal_ble_gap_connect()` ?

Reply:
E.g. BLE pairing request. It may be generated pretty fast if peer device send the request on connected. I have experienced this many times during the fixture test.

Reply:
Okay, in that case I don't see a way of resolving this unless we switch over to a CONNECTED event to fill in the necessary data in wiring directly from BLE stack events. So, let's add it. Perhaps to avoid breaking Gen 3 stuff, add a feature flag for now for rtl872x platforms specifically and have this behavior only on those for now until we have a bit more time.

Reply:
Will do.